### PR TITLE
AL Income documentation

### PR DESCRIPTION
**Pull request by BryceStevenWilley**
Still in progress, but you can see the general shape of how I plan on writing the documentation pages in the ALExpenseList page. 

Polish still needed on the ALExpenseList page:
* doing a quick run through of the tutorial myself as a newer DA user

Additions needed for the rest of this PR in order of priority:
* ALItemizedJobList
* ALAssetList
* ALIncomeList

### CODE REVIEW

**Comment by BryceStevenWilley** [comment]
@nonprofittechy and @plocket, I've added you as reviewers for the moment to make sure that the general approach of a tutorial like this would be useful and get any ideas y'all might have to improve said tutorial. I'll ping you both again when I've written all of the tutorials for ALIncome and change this from a draft to a ready PR then.

File: docs/alincome/expenses.md
Comment (plocket) [inline docs/alincome/expenses.md:10]: I've noticed you've used the word 'simple' a couple times. You were trying to avoid that, right?

File: docs/alincome/expenses.md
Comment (plocket) [inline docs/alincome/expenses.md:17]:
```suggestion
don't want the user forget to list any significant expenses they would have
```
To match the previous part of the sentence.

File: docs/alincome/expenses.md
Comment (plocket) [inline docs/alincome/expenses.md:50]:
```suggestion
* have the [`ALToolbox` or `AssemblyLine` package installed on your server](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/installation#run-the-installation-script)
```
Many people will probably just be using AssemblyLine.

File: docs/alincome/expenses.md
Comment (plocket) [inline docs/alincome/expenses.md:191]: I thought I remembered that multiple expenses could have the same source? If so, "Value of just one category of expense" and the description for "source" might need to be changed too. If not, never mind!

File: docs/alincome/expenses.md
Comment (BryceStevenWilley) [reply docs/alincome/expenses.md:10]: Good catch. I was trying to avoid using it to imply that some coding process was simple, like "Simply change the docassemble configuration or something".  Making things simple (and straight-forward) for our users is an important goal of our libraries though, so it feels valid here? Open to suggestions, but can think of any improvements myself that don't make this sentence longer than it needs to be.

File: docs/alincome/expenses.md
Comment (BryceStevenWilley) [reply docs/alincome/expenses.md:50]: Then they'll have `ALToolbox` already installed? I think I understand what you're pointing out; ppl might not immediately recognize the toolbox package, but they would know if they've installed the general assembly line. I might rephrase this differently to address that.

```suggestion
* have the [`AssemblyLine` package installed on your server](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/installation#run-the-installation-script), or have installed the [`ALToolbox` package](https://github.com/SuffolkLITLab/docassemble-ALToolbox) separately.
```

File: docs/alincome/expenses.md
Comment (plocket) [reply docs/alincome/expenses.md:10]: "Simple" here seems a bit over confident, but I see what you're saying. I thought I caught a use of the word lower down being used to express that something was simple to do. I'll have to try and track that down.

```suggestion
The `ALExpenseList` is a class that helps you ask a user about their different expenses.
```

File: docs/alincome/jobs.md
Comment (plocket) [inline docs/alincome/jobs.md:34]: Maybe this just needs a different word. Do you mean simple as in easy for a user to use, simple for the developer to use, or simple as in bare bones?

File: docs/alincome/jobs.md
Comment (plocket) [inline docs/alincome/jobs.md:10]: Simple for the developer and end user or just for the end user? e.g.
* "...but simple for you to use and simple for your users to answer..."?
* "...but simple for your users to complete..."?

File: docs/alincome/expenses.md
Comment (BryceStevenWilley) [reply docs/alincome/expenses.md:191]: Yes and no. Yes, `ALExpenseList` can have more than once expense from the same source. However the UX that we've landed on as the default (and is being used in this tutorial) won't have things from the same source, as it's showing a dict of checkboxes before hand, with each key for the checkbox as one source.

A user can select other and then add another expense from the same source, but TBH, that was more for resiliency and couldn't remember if there was a specific feature that someone would want to use with that.

Trying to balance being accurate and not confusing users with technical details they shouldn't have to know about to use the library is hard.  The description of source is accurate, and I don't think I would change it, even if we talk about allowing duplicates. I might change this line to be a bit more accurate though.

```suggestion
  Value of expenses with the source "rent": ${ currency(expenses.total(source="rent")) }
```

File: docs/alincome/jobs.md
Comment (BryceStevenWilley) [reply docs/alincome/jobs.md:34]: Simple as in bare bones in this case. This is definitely a sub-par usage of the word, thanks.

File: docs/alincome/jobs.md
Comment (BryceStevenWilley) [reply docs/alincome/jobs.md:10]: Simple for both, but this is another bad usage of the term. I'm being outed as adding the very thing I'm trying to remove, lol

**Comment by plocket** [comment]
Is it worth adding a little bit to ALJobs/ALJobsList that mentions ALItemizedJob? In a way that describes the difference, so folks know have more context about when to use which one.

**Comment by BryceStevenWilley** [comment]
I do already: https://github.com/SuffolkLITLab/docassemble-AssemblyLine-documentation/pull/310/files#diff-0d907de082071684a35b5c03fcc8804ddbc8fc896ebe76e0fb81489254f91428R80-R81. I plan on having a few more pointers to Itemized Jobs from Jobs, and a fuller comparison in `ALItemizedJob` though (not in `ALJob`, just to reduce up-front mental burden).

**Comment by BryceStevenWilley** [comment]
@plocket , this is ready for review, I know you've been looking at it already, but feel free to do so again, I should have addressed your comments so far.

Not pinging nonprofittechy, since he's focused on other things this week.

File: docs/alincome/itemizedjobs.md
Comment (plocket) [inline docs/alincome/itemizedjobs.md:19]: Caroline has said that the main reason she wants this is because otherwise the user doesn't keep track of everything they need to take into account and they have to do a lot of the math themselves. You say a bit about that at the bottom of this section, but maybe it's worth thinking about which of those is the main purpose. For example, I think I understood that Caroline preferred it as a user experience over ALJob.

File: docs/alincome/itemizedjobs.md
Comment (plocket) [inline docs/alincome/itemizedjobs.md:38]: In aim of the above, maybe something like this would give fuller context:

```suggestion
* have looked at the [the `ALJob` tutorial](jobs.md) and decided you want to do the user's calculations for them or you need to handle more complicated information
```

File: docs/alincome/overview.md
Comment (plocket) [inline docs/alincome/overview.md:22]: I like that we have a demo up. Maybe list that first, since people (including me) often jump to do things instead of reading through to the end. Unless there's an advantage to them doing it instead.

File: docs/alincome/overview.md
Comment (plocket) [inline docs/alincome/overview.md:32]: 👍 I like this temporary alternative.

**Comment by plocket** [comment]
Did you get to test all the code you showed as it's shown in the docs? Some of this is, I think, taken from parts of the demo interview, but have these pieces of code been independently tested? We could add them as hidden code to the demo interview and maybe just trigger the blocks when we need to make adjustments or for testing or something so we can know when things fall behind or if we've left something out. That's a heck of a lot of work, though. Maybe that could be a future issue.

**Review by plocket — COMMENTED** [review]
I've not re-read the whole thing and I'm not sure what tomorrow will bring. For now, I'll just comment that this has been a good chunk of work and is at the very least a solid start. I think it's probably acceptable to merge. If I get to go through it again, I'll try to give it a more sure yay or nay.

**Review by plocket — APPROVED** [review]
Don't know what I'm on about! Take my other comments as you want. Approved unless I find something terribly wrong.

File: docs/alincome/itemizedjobs.md
Comment (BryceStevenWilley) [reply docs/alincome/itemizedjobs.md:19]: Hm, if that's the case, there isn't a strong reason for ALJob at all? Should we still even document it / suggest it, or should we steer everyone towards ItemizedJob? 

I see what she means, but TBH I'm not sure if ALItemizedJob is there yet in terms of UX. I know that's the purpose of a lot of our other classes, like ALExpenseList, where we can ask the user about a lot of different expenses so they don't have to keep track of them. But ItemizedJob still feels cumbersome when doing that, and it doesn't quite remove the need for users to make calculations (a lot more discussion of that in https://github.com/SuffolkLITLab/docassemble-ALToolbox/issues/140).

Comment (BryceStevenWilley) [reply docs/alincome/itemizedjobs.md:19]: Now that I've thought about https://github.com/SuffolkLITLab/docassemble-ALToolbox/issues/140, I'm realizing that the core functionality of ALItemizedJob of handling additional income like tips, commissions, and overtime wages isn't flexible enough; all of those additional incomes are inconsistent, and asking about them in aggregate either require the user to do math (i.e. directly asking "what's your usual amount" or "what's the average"), or they just get over or under estimated.

I did a quick "look at this website I made" user test with my partner, and she pointed out some of these inconsistencies and unclear instructions. I'll make an issue for it separately, but I do think that the current language should probably stay for now, and maybe I'll add a warning with the above info as well.

Also will directly invite @CaroRob to weigh in if she has any thoughts.

**Comment by BryceStevenWilley** [comment]
> Did you get to test all the code you showed as it's shown in the docs?

Yes! I'll double test it before merging, in case I changed anything.

> We could add them as hidden code to the demo interview and maybe just trigger the blocks when we need to make adjustments or for testing or something so we can know when things fall behind or if we've left something out

I like the idea of automatically testing documentation, but if we do that I want a way to directly include the tested code into the documentation, which I don't know how to do right now. I'll make an issue for it.

**Comment by BryceStevenWilley** [comment]
Waiting for https://github.com/SuffolkLITLab/docassemble-ALToolbox/pull/145 to be merged before merging here.

**Pull request merged**
